src/editor/controls/FlyControls.js

```diff
@@ -1,57 +1,104 @@
 import THREE from "../../vendor/three";
 
+const EPSILON = 10e-5;
+const UP = new THREE.Vector3(0, 1, 0);
+const parentInverse = new THREE.Matrix4();
+const IDENTITY = new THREE.Matrix4().identity();
+const vec = new THREE.Vector3();
+const quat = new THREE.Quaternion();
+const worldPos = new THREE.Vector3();
+const worldQuat = new THREE.Quaternion();
+const worldScale = new THREE.Vector3();
+const candidateWorldQuat = new THREE.Quaternion();
+
 export default class FlyControls {
   constructor(camera, inputManager) {
     this.enabled = false;
     this.camera = camera;
     this.inputManager = inputManager;
     this.moveSpeed = 2;
     this.boostSpeed = 2;
-    this.lookSensitivity = 1;
+    this.lookSensitivity = 10;
+    this.maxXRotation = THREE.Math.degToRad(80);
     this.direction = new THREE.Vector3();
-    this.maxXRotation = THREE.Math.degToRad(70);
-  }
 
-  onClickCanvas = () => {
-    if (this.enabled && document.pointerLockElement !== this.inputManager.canvas) {
-      this.inputManager.canvas.requestPointerLock();
-    }
-  };
+    this.onMouseUp = e => {
+      if (this.enabled && e.button === 2) {
+        if (document.pointerLockElement === this.inputManager.canvas) {
+          document.exitPointerLock();
+        }
+      }
+    };
+  }
 
   enable() {
     this.enabled = true;
+    this.inputManager.canvas.requestPointerLock();
     this.inputManager.enableInputMapping(Fly, FlyMapping);
+    this.inputManager.canvas.addEventListener("mouseup", this.onMouseUp);
   }
 
   disable() {
     this.enabled = false;
     this.inputManager.disableInputMapping(Fly);
+    this.inputManager.canvas.removeEventListener("mouseup", this.onMouseUp);
   }
 
   update(dt) {
     if (!this.enabled) return;
-
     const input = this.inputManager;
 
-    const rotation = this.camera.rotation;
-    const lookSensitivity = this.lookSensitivity;
+    // assume that this.camera[position,quaterion/rotation,scale] are authority
+    this.camera.updateMatrix();
+    this.camera.updateMatrixWorld();
+    this.camera.matrixWorld.decompose(worldPos, worldQuat, worldScale);
+
+    // rotate about the camera's local x axis
+    candidateWorldQuat.multiplyQuaternions(
+      quat.setFromAxisAngle(vec.set(1, 0, 0).applyQuaternion(worldQuat), input.get(Fly.lookY) * this.lookSensitivity),
+      worldQuat
+    );
 
-    let xRotation = rotation.x > Math.PI ? rotation.x - 2 * Math.PI : rotation.x;
-    xRotation = THREE.Math.clamp(xRotation, -this.maxXRotation, this.maxXRotation);
+    // check change of local "forward" and "up" to disallow flipping
+    const camUpY = vec.set(0, 1, 0).applyQuaternion(worldQuat).y;
+    const newCamUpY = vec.set(0, 1, 0).applyQuaternion(candidateWorldQuat).y;
+    const newCamForwardY = vec.set(0, 0, -1).applyQuaternion(candidateWorldQuat).y;
+    const extrema = Math.sin(this.maxXRotation);
+    const allowRotationInX =
+      newCamUpY > 0 && ((newCamForwardY < extrema && newCamForwardY > -extrema) || newCamUpY > camUpY);
```

Comment:
If we only want to disallow flipping, use
`const allowRotationInX = newCamUpY > 0`
This instead clamps the rotation when it can, or if it is out of the clamp range (e.g. orbit controls left you looking straight down) it only allows rotation back towards the clamped range.